src/main/java/com/conveyal/datatools/manager/jobs/ProcessSingleFeedJob.java

```diff
@@ -48,21 +65,102 @@ public String getFeedSourceId () {
+                ArbitraryTransformJob zipTransform = new ArbitraryTransformJob(owner, feedVersion.retrieveGtfsFile(), transformation);
+                // Run transform job in line so we can monitor the error status before load/validate begins.
+                zipTransform.run();
+                // Short circuit the feed load/validate if a pre-load transform fails.
+                if (zipTransform.status.error) return;
+            }
+        }
 
         // First, load the feed into database. During this stage, the GTFS file will be uploaded to S3 (and deleted locally).
         addNextJob(new LoadFeedJob(feedVersion, owner, isNewVersion));
 
         // Next, validate the feed.
         addNextJob(new ValidateFeedJob(feedVersion, owner, isNewVersion));
 
+        // We only need to snapshot the feed if there are transformations at the database level. In the case that there
+        // are, the snapshot namespace will be the target of these modifications. If we were to apply the modifications
+        // to the original feed version's namespace, the namespace and the zip file would become out of sync, which
+        // violates a core expectation for this application: that each zip file represents exactly its database
+        // representation.
+        boolean shouldSnapshot = shouldTransform && rules.hasTransformationsOfType(feedVersion, DbTransformation.class);
+        if (shouldSnapshot) {
+            Snapshot snapshot = new Snapshot("Transform of " + feedVersion.name, feedVersion);
+            addNextJob(new CreateSnapshotJob(owner, snapshot));
+            // Apply post-load transformations to snapshotted feed version. Post-load transformations will modify only the
+            // snapshot (not the original feed version's namespace), so the snapshot must be published (or loaded into the
+            // editor) in order to see the results.
+            // FIXME need to add snapshot namespace...
+            for (DbTransformation transformation : rules.getActiveTransformations(feedVersion, DbTransformation.class)) {
+                addNextJob(new ArbitraryTransformJob(owner, snapshot.id, transformation));
+            }
+            // If the user has selected to create a new version from the resulting snapshot, do so here.
+            if (rules.createNewVersion) {
+                addNextJob(new CreateFeedVersionFromSnapshotJob(feedSource, snapshot, owner));
+            }
+        }
+
+        // FIXME: Should we overwrite the input GTFS dataset if transforming in place?
+
         // TODO: Any other activities that need to be run (e.g., module-specific activities).
+        if (DataManager.isModuleEnabled("deployment")) {
+//            Project project = feedSource.retrieveProject();
+//            if (project.autoDeployId)
+            // TODO: Get deployment, update feed version for feed source, and kick off deployment to server that
+            //  deployment is currently pointed at.
```

Comment:
This seems like it needs more thought than just a TODO. In the case of a project with many feed sources, you might want to run something that updates all feed sources and then deploys as needed.

Reply:
What would you like to see instead of the TODO? A GitHub issue perhaps? This hasn't been built, but was intended just as a placeholder for an idea I had.

Reply:
Yeah, I'd remove this comment and make an issue. Automated deployments isn't even a part of the MTC wishlist in https://github.com/ibi-group/datatools-ui/issues/544.